### PR TITLE
Fix bug 1225302: Add filter by translation author

### PR DESCRIPTION
**Pull request by mathjazz**
@jotes r?

Authors are listed in the filter menu. Max height is set to fit 3 authors, then you need to scroll. If there aren't any translations submitted by Pontoon users, we don't show the filter. We update the authors list on translate page load, translation submission/deletion, batch operations and file upload.

I haven't found a much better UX solution for the authors selector. Filter menu is getting huge. Also, we'll be adding date/time (bug 1225305) and errors/checks (bug 1237667) filters, which will make it even bigger.

We should probably refactor filter UI as part of 1243115. We might also be able to merge it with the progress bar menu.

### CODE REVIEW

File: pontoon/base/models.py [hunk 1]
```diff
@@ -1148,6 +1158,9 @@ def for_project_locale(self, project, locale, paths=None, filter_type=None,
             elif filter_type == 'unchanged':
                 entities = self.objects.unchanged(locale)
 
+            elif filter_type in Translation.authors(locale, project, paths).values_list('email', flat=True):
```
Comment (jotes) [inline pontoon/base/models.py]: What would you say to change this DB-call for the regexp email matching? Could be a little bit faster.

File: pontoon/base/models.py [hunk 2]
```diff
@@ -1309,6 +1317,28 @@ class Translation(DirtyFieldsMixin, models.Model):
...
+
+        if paths:
+            paths = project.parts_to_paths(paths)
+            translations = translations.filter(entity__resource__path__in=paths)
+
+        authors = User.objects.filter(translation__in=translations).distinct()
+
+        if serialize:
+            authors_list = []
+            for a in Translation.authors(locale, project, paths):
```
Comment (jotes) [inline pontoon/base/models.py]: Isn't `Translation.authors(locale, project, paths)` the same as `authors`?

File: pontoon/base/models.py [hunk 3]
```diff
@@ -1309,6 +1317,28 @@ class Translation(DirtyFieldsMixin, models.Model):
...
+    def authors(self, locale, project, paths, serialize=False):
+        translations = Translation.objects.filter(entity__resource__project=project, locale=locale)
+
+        if paths:
+            paths = project.parts_to_paths(paths)
+            translations = translations.filter(entity__resource__path__in=paths)
+
+        authors = User.objects.filter(translation__in=translations).distinct()
+
+        if serialize:
```
Comment (jotes) [inline pontoon/base/models.py]: I feel that we shouldn't mix up data sources with serialization methods

File: pontoon/base/static/js/translate.js [hunk 4]
```diff
@@ -1953,6 +1964,27 @@ var Pontoon = (function (my) {
...
+      $('#filter').find('.authors, .for-authors').toggle(this.authors.length > 0);
+      var authors = $('#filter .authors ul').empty();
+
+      $.each(this.authors, function() {
+        authors.append('<li data-type="' + this.email + '">' +
+          '<figure class="author">' +
+            '<img class="rounded" src="' + this.gravatar_url + '">' +
+            '<figcaption>' +
+              '<p class="name">' + this.display_name + '</p>' +
+              '<p class="email">' + this.email + '</p>' +
```
Comment (jotes) [inline pontoon/base/static/js/translate.js]: I know It's a bigger topic but maybe we should change .email into username to protect contributors privacy? At some point of course.

**Comment by jotes** [comment]
r+ :-)

File: pontoon/base/models.py (hunk 1 above)
Comment (mathjazz) [inline pontoon/base/models.py]: Yes, that's a possible optimization. It's also slightly less powerful than this check. For now I'll leave it, but it's good to know we can be faster here.

File: pontoon/base/models.py (hunk 2 above)
Comment (mathjazz) [inline pontoon/base/models.py]: Ouch!

File: pontoon/base/static/js/translate.js (hunk 4 above)
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Yeah. I like the idea users don't need anything besides their email address to authenticate and be presented, but I also understand the privacy concern.

**Comment by mathjazz** [comment]
Updated. Also enabled author filter in URLs.

**Comment by jotes** [comment]
r+ x2

**Pull request merged**